### PR TITLE
feat: implement custom TES authorization.

### PR DESCRIPTION
**Pull request by peterhuene**
This PR adds the `Authorizer` trait to the crate which can be used when constructing a `Client`.

An authorizer is responsible for modifying requests that the TES client sends to the TES service to include an `Authorization` header.

Three default implementations of the `Authorizer` trait are included:

* An authorizer for basic (username and password).
* An authorizer for bearer token.
* An authorizer for OAuth device code flow.

The OAuth authorizer takes a required "prompt" callback to display the verification URI and user code to the user.

An optional reauthorization callback can be used with the OAuth authorizer for handling token refresh errors or prevent the authorizer from attempting device authorization again.

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added tests (when appropriate).
- [x] You have updated the README or other documentation to account for these
      changes (when appropriate).
- [x] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.1.0/

### CODE REVIEW

File: src/v1/client.rs [hunk 1]
```diff
@@ -341,4 +323,139 @@ impl Client {
...
+        let mut initial = true;
+
+        loop {
+            // Create the request and apply the authorization
+            let mut req = request();
+            if let Some(authorizer) = &self.authorizer {
+                req = authorizer
+                    .authorize(initial, req)
+                    .await
+                    .map_err(Error::Authorization)?;
```
Comment (claymcleod) [inline src/v1/client.rs]: Authorization failures come out of here as *transient* retry errors, so the retry strategy re-runs authorization on each attempt. That's because `impl<E> From<E> for RetryError<E>` in tokio-retry2 builds `RetryError::Transient`. If the user denies the device code, or lets it expire, each retry starts a new device flow and prompts again. If `on_reauthorization` returns `false`, the next attempt resends the stale token, gets a 401, and asks the handler again, so the user's "no" doesn't stick. Let's make these permanent. If you want network failures reaching the token endpoint to stay retryable, classify only `oauth::Error::Reqwest` as transient.

```suggestion
                    .map_err(|e| RetryError::permanent(Error::Authorization(e)))?;
```

File: src/auth.rs [hunk 2]
```diff
@@ -0,0 +1,791 @@
...
+        if initial && let Some(tokens) = tokens.as_ref() {
+            return Ok((
+                tokens.refresh.is_some(),
+                request.map(|r| r.bearer_auth(tokens.access.secret())),
+            ));
+        }
+
+        // If this is not the initial request, attempt to refresh the access
+        // token if it possible to do
+        if !initial {
```
Comment (claymcleod) [inline src/auth.rs]: Concurrent requests that all get a 401 each refresh again, or prompt for a new device code, one after another. Nothing checks whether an earlier caller already replaced the token. I think `authorize` needs to know which token was rejected, either the token itself or a generation counter. Then this branch can reuse the current token when it has already changed.

File: src/auth.rs [hunk 3]
```diff
@@ -0,0 +1,791 @@
...
+            if let Some(refresh) = tokens.as_ref().and_then(|tokens| tokens.refresh.as_ref()) {
+                // Refresh the token
+                match oauth::refresh_token(&self.config, refresh).await {
+                    Ok((access, refresh)) => {
+                        let has_refresh = refresh.is_some();
```
Comment (claymcleod) [inline src/auth.rs]: When the refresh response has no `refresh_token`, this throws away the refresh token we already have. Servers are allowed to omit it (RFC 6749 §6), and then the next 401 falls back to a device prompt. Consider keeping the old one:

```suggestion
                    Ok((access, new_refresh)) => {
                        let refresh = new_refresh.or_else(|| Some(refresh.clone()));
                        let has_refresh = refresh.is_some();
```

File: src/auth.rs [hunk 4]
```diff
@@ -0,0 +1,791 @@
...
+                        request = request.map(|r| r.bearer_auth(access.secret()));
+                        *tokens = Some(Tokens { access, refresh });
+                        return Ok((has_refresh, request));
+                    }
+                    Err(e) => {
```
Comment (claymcleod) [inline src/auth.rs]: Every refresh failure ends up here, including network errors and 5xx responses. So a short outage at the token endpoint drops the refresh token and starts a device flow. It might be worth falling back only when the server rejects the grant (a `RequestTokenError::ServerResponse`), and returning transport errors as they are.

File: src/v1/client.rs [hunk 5]
```diff
@@ -48,31 +55,31 @@ pub enum Error {
...
     /// An error when serializing or deserializing JSON.
     #[error(transparent)]
     SerdeParams(#[from] serde_url_params::Error),
-
     /// An error from `reqwest`.
     #[error(transparent)]
     Reqwest(#[from] reqwest::Error),
+    /// An error occurred during authorization.
+    #[error("authorization with the TES service failed: {0}")]
+    Authorization(crate::auth::Error),
```
Comment (claymcleod) [inline src/v1/client.rs]: Without `#[source]`, the cause chain stops here, so users see "error response from `…`" but never the server's error. Let's mark the source so error reporters like miette and anyhow can print the whole chain:

```suggestion
    #[error("authorization with the TES service failed")]
    Authorization(#[source] crate::auth::Error),
```

File: src/oauth.rs [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+        .set_device_authorization_url(DeviceAuthorizationUrl::from_url(
+            config.authorization.clone(),
+        ))
+        .set_token_uri(TokenUrl::from_url(config.token.clone()));
+
+    if let Some(secret) = &config.client_secret {
+        client = client.set_client_secret(ClientSecret::new(secret.clone()));
+    }
+
+    let http_client = reqwest::ClientBuilder::new()
```
Comment (claymcleod) [inline src/oauth.rs]: This client has no timeout, and the token lock is held for the whole flow. A hung device or token endpoint blocks every request on the `Client` indefinitely. Let's set `.timeout(..)` and `.connect_timeout(..)` here and in `refresh_token`. Building the client once and passing it in would also let both functions reuse connections.

File: Cargo.toml [hunk 7]
```diff
@@ -29,18 +30,21 @@ types = ["dep:url"]
 
 [dependencies]
 chrono = { version = "0.4.41", features = ["serde"] }
+futures = "0.3.34"
+oauth2 = { version = "5.0.0", optional = true }
```
Comment (claymcleod) [inline Cargo.toml]: `oauth2`'s default features pull in reqwest 0.12 with bundled webpki roots, next to the crate's reqwest 0.13, which uses the platform verifier. So every `client` build compiles two HTTP/TLS stacks. OAuth calls trust a different set of CAs than TES calls, which breaks identity providers behind an internal CA. And `oauth::Error::Reqwest` exposes a reqwest 0.12 type. Consider `default-features = false` and passing oauth2 a closure over our own reqwest 0.13 client. oauth2 implements `AsyncHttpClient` for any `Fn(HttpRequest) -> Future`.

File: Cargo.toml [hunk 8]
```diff
@@ -29,18 +30,21 @@ types = ["dep:url"]
 
 [dependencies]
 chrono = { version = "0.4.41", features = ["serde"] }
+futures = "0.3.34"
```
Comment (claymcleod) [inline Cargo.toml]: `futures` and `tokio` are only used by the `client`-gated modules, but they aren't optional, so `types`-only builds now pull in both. Let's mark them `optional = true` and add `"dep:futures"` and `"dep:tokio"` to the `client` feature.

File: src/v1/client/builder.rs [hunk 9]
```diff
@@ -28,19 +30,18 @@ pub enum Error {
 pub type Result<T> = std::result::Result<T, Error>;
 
 /// A builder for a [`Client`](Client).
-#[derive(Clone, Debug, Default)]
+#[derive(Default)]
```
Comment (claymcleod) [inline src/v1/client/builder.rs]: Dropping `Clone` here is a breaking change for anyone who clones a `Builder`. It's only needed because `Box<dyn Authorizer>` isn't `Clone`. One option is to store `Option<Arc<dyn Authorizer>>` and add blanket `Authorizer` impls for `Box<T>` and `Arc<T>`. That keeps the derive, lets clients share one OAuth login, and makes the `_boxed` setters unnecessary.

File: examples/service-info.rs [hunk 10]
```diff
@@ -33,26 +33,25 @@ async fn main() -> Result<()> {
...
+    if username.is_none() && password.is_some() {
         panic!("${USER_ENV} and ${PASSWORD_ENV} must both be set to use basic auth");
     }
 
-    if let Some(username) = username {
-        let credentials = format!("{}:{}", username, password.unwrap());
-        let encoded = BASE64_STANDARD.encode(credentials);
-        builder = builder.insert_header("Authorization", format!("Basic {encoded}"));
-    }
+    let authorizer = username.map(|username| BasicAuthorizer::new(username, password));
```
Comment (claymcleod) [inline examples/service-info.rs]: `$USER` is set by the shell on almost every Unix machine, so with this change the examples send `Basic <login name>:` whenever `$PASSWORD` is unset. Previously they failed instead. The panic message above also still says both variables must be set. Let's rename the variables to something like `TES_USER` and `TES_PASSWORD` across the examples.

File: src/auth.rs [hunk 11]
```diff
@@ -0,0 +1,791 @@
...
+        request.send().await.unwrap();
+
+        endpoint.assert();
+    }
+
+    #[tokio::test]
+    async fn test_oauth_authorizer() {
+        let mut oauth = OAuthTestServer::new(true).await;
+
+        let invoked = Arc::new(AtomicBool::new(false));
```
Comment (claymcleod) [inline src/auth.rs]: `invoked` is set in every OAuth test here but never asserted, so nothing checks that the prompt ran or how often. Consider making it an `AtomicUsize` counter and asserting the expected count. It would also be worth adding tests for a reauth handler that returns `true`, an authorization failure with non-empty retries, and concurrent 401s.

File: CHANGELOG.md [hunk 12]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Added
+
+- Added support for OAuth authorization for the TES client ([#29](https://github.com/stjude-rust-labs/tes/pull/29)).
```
Comment (claymcleod) [inline CHANGELOG.md]: The entry covers OAuth, but not the `Authorizer` trait, the Basic and Bearer authorizers, or the new builder methods. It also skips the breaking changes: `Builder` no longer implements `Clone`, `client::Error` has a new variant (it's exhaustive, so downstream matches break), and the MSRV is now 1.88.0. Let's add those under "Added" and "Changed" so the next release is clearly 0.11.0.

File: src/auth.rs [hunk 13]
```diff
@@ -0,0 +1,791 @@
...
+         access token"
+    )]
+    NoRefreshToken,
+}
+
+/// Represents a generic authorizer of TES requests.
+pub trait Authorizer: Send + Sync {
+    /// Authorizes for the given request.
+    ///
+    /// Upon the first request by the client, `initial` will be `true`.
```
Comment (claymcleod) [inline src/auth.rs]: `initial` is `true` on the first attempt of every request, not only the client's first request:

```suggestion
    /// On the first attempt of each request, `initial` will be `true`.
```

File: src/auth.rs [hunk 14]
```diff
@@ -0,0 +1,791 @@
...
+        &'a self,
+        _initial: bool,
+        request: RequestBuilder,
+    ) -> BoxFuture<'a, Result<RequestBuilder, Error>> {
+        ready(Ok(request.bearer_auth(&self.0))).boxed()
+    }
+}
+
+/// Collection of OAuth-related tokens.
+struct Tokens {
```
Comment (claymcleod) [inline src/auth.rs]: Since `expires_in` is dropped, a refresh happens only after a 401. That costs a rejected round trip after each expiry, and it never happens if a service answers with 403 instead. It might be worth storing an `expires_at: Option<Instant>` here and refreshing ahead of time, with a little slack, keeping the 401 path as the fallback.

File: src/oauth.rs [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+            Error::Basic(
+                BasicError {
+                    uri: config.token.clone(),
+                    error,
+                }
+                .into(),
+            )
+        })?;
+
+    Ok((
```
Comment (claymcleod) [inline src/oauth.rs]: `authorize_device` rejects non-bearer tokens, but this path doesn't check `token_type`, so a refreshed non-bearer token would still be sent as `Bearer`. Consider pulling the check and the token-pair conversion into a shared helper that both functions use.

File: src/auth.rs [hunk 16]
```diff
@@ -0,0 +1,791 @@
...
+pub enum Error {
+    /// An OAuth error occurred.
+    #[error(transparent)]
+    Oauth(#[from] oauth::Error),
+    /// No refresh token was provided the client.
+    #[error(
+        "the client did not receive a refresh token and could not automatically refresh the \
+         access token"
+    )]
+    NoRefreshToken,
```
Comment (claymcleod) [inline src/auth.rs]: This is only returned when the `on_reauthorization` handler declines, but the name and message read as if a refresh failed. Consider something like `ReauthorizationDeclined`, with a message saying reauthorization was declined because no refresh token was available.

File: src/auth.rs [hunk 17]
```diff
@@ -0,0 +1,791 @@
...
+    /// client or if a token refresh operation failed.
+    ///
+    /// The provided error will be `Some` if the token refresh operation failed.
+    ///
+    /// If the handler returns `true`, a new device code authorization flow is
+    /// attempted.
+    ///
+    /// If the handler returns `false`, an error is returned as the result of
+    /// authorization.
+    pub fn on_reauthorization<E>(mut self, reauth: E) -> Self
```
Comment (claymcleod) [inline src/auth.rs]: I think the natural use of this handler is asking the user whether to log in again, and a sync callback would block a Tokio worker while holding the token lock. It might be worth making it return a future, or at least documenting that it must not block, before this ships and becomes hard to change.

File: src/lib.rs [hunk 18]
```diff
@@ -17,4 +17,10 @@
 //!     https://editor.swagger.io/?url=https://ga4gh.github.io/task-execution-schemas/openapi.yaml
 #![doc = include_str!("../docs/FEATURES.md")]
 
+#[cfg(feature = "client")]
+pub mod oauth;
```
Comment (claymcleod) [inline src/lib.rs]: One option is to move this under `auth` as `auth::oauth` and gate it on its own `oauth` feature. Basic and Bearer users then wouldn't pull in `oauth2`, and an `oauth2` major bump wouldn't force a breaking release for them. `docs/FEATURES.md` would need a row for it.

File: src/v1/client.rs [hunk 19]
```diff
@@ -1,7 +1,12 @@
 //! A client for interacting with a Task Execution Service (TES) service.
 
+use std::fmt;
 use std::time::Duration;
 
+use reqwest::Method;
+pub use reqwest::RequestBuilder;
```
Comment (claymcleod) [inline src/v1/client.rs]: `RequestBuilder` is what `Authorizer` implementers need, so it might fit better re-exported from `tes::auth`, next to the trait, with a private import here.

File: src/auth.rs [hunk 20]
```diff
@@ -0,0 +1,791 @@
...
+    /// Upon success, returns a tuple of whether or not there is a refresh token
+    /// and the provided request.
+    async fn perform_authorization(
+        &self,
+        initial: bool,
+        mut request: Option<RequestBuilder>,
+    ) -> Result<(bool, Option<RequestBuilder>), Error> {
+        // Take a lock on the tokens for the entire Oauth operation.
+        // This is intentionally a long-lived lock as at most one
+        // authorization flow should occur
```
Comment (claymcleod) [inline src/auth.rs]:
```suggestion
        // Take a lock on the tokens for the entire OAuth operation.
        // This is intentionally a long-lived lock as at most one
        // authorization flow should occur at a time.
```

File: src/auth.rs [hunk 21]
```diff
@@ -0,0 +1,791 @@
...
+        let mut tokens = self.tokens.lock().await;
+
+        // If this is the initial request and we have an access token, use
```
Comment (claymcleod) [inline src/auth.rs]:
```suggestion
        // If this is the initial request and we have an access token, use it.
```

File: src/auth.rs [hunk 22]
```diff
@@ -0,0 +1,791 @@
...
+        if initial && let Some(tokens) = tokens.as_ref() {
+            return Ok((
+                tokens.refresh.is_some(),
+                request.map(|r| r.bearer_auth(tokens.access.secret())),
+            ));
+        }
+
+        // If this is not the initial request, attempt to refresh the access
+        // token if it possible to do
```
Comment (claymcleod) [inline src/auth.rs]:
```suggestion
        // If this is not the initial request, attempt to refresh the access
        // token if it is possible to do so.
```

File: src/auth.rs [hunk 23]
```diff
@@ -0,0 +1,791 @@
...
+use tokio::sync::Mutex;
+
+use crate::oauth;
+
+/// Represents an authorization error.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// An OAuth error occurred.
+    #[error(transparent)]
+    Oauth(#[from] oauth::Error),
```
Comment (claymcleod) [inline src/auth.rs]: Let's match the `OAuthAuthorizer` casing while this is still unreleased (the test near the end of the file needs the same rename):

```suggestion
    OAuth(#[from] oauth::Error),
```

File: src/oauth.rs [hunk 24]
```diff
@@ -0,0 +1,526 @@
...
+#[error("error response from `{uri}`")]
+pub struct BasicError {
+    /// The URI of the request.
+    pub uri: Url,
+    /// The response error.
+    #[source]
+    pub error: BasicResponseError,
+}
+
+/// A basic error from an OAuth request.
```
Comment (claymcleod) [inline src/oauth.rs]:
```suggestion
/// An error from an OAuth device access token request.
```

File: examples/status-table.rs [hunk 25]
```diff
@@ -149,28 +148,24 @@ async fn main() -> Result<()> {
...
-        .into_diagnostic()
-        .context("URL could not be parsed")?;
-
     let username = std::env::var(USER_ENV).ok();
     let password = std::env::var(PASSWORD_ENV).ok();
 
-    if (username.is_some() && password.is_none()) || (username.is_none() && password.is_some()) {
-        bail!("${USER_ENV} and ${PASSWORD_ENV} must both be set to use basic auth");
+    if username.is_none() && password.is_some() {
+        panic!("${USER_ENV} and ${PASSWORD_ENV} must both be set to use basic auth");
```
Comment (claymcleod) [inline examples/status-table.rs]: This was `bail!` before, and `main` returns `miette::Result`, so let's keep returning an error here (and in the other examples). That also needs `use miette::bail;` back:

```suggestion
        bail!("${PASSWORD_ENV} requires ${USER_ENV} to be set to use basic auth");
```

File: src/v1/client.rs [hunk 26]
```diff
@@ -341,4 +323,139 @@ impl Client {
...
+            }
+
+            // Attempt reauthorization if needed; this occurs at most once
+            if initial
+                && response.status() == StatusCode::UNAUTHORIZED
+                && self
+                    .authorizer
+                    .as_ref()
+                    .map(|a| a.reauthorizes())
+                    .unwrap_or(false)
```
Comment (claymcleod) [inline src/v1/client.rs]:
```suggestion
                && self.authorizer.as_ref().is_some_and(|a| a.reauthorizes())
```

File: src/v1/client.rs [hunk 27]
```diff
@@ -341,4 +323,139 @@ impl Client {
...
+        }
+    }
+}
+
+impl fmt::Debug for Client {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Client")
+            .field("url", &self.url)
+            .field("client", &self.client)
+            .field("authorizer", &"...")
```
Comment (claymcleod) [inline src/v1/client.rs]: This prints `"..."` even when no authorizer is set. Printing whether one is present keeps the secret hidden and is still useful (same in `Builder`'s impl):

```suggestion
            .field("authorizer", &self.authorizer.is_some())
```

**Review by claymcleod — CHANGES_REQUESTED** [review]
An OAuth device-flow example would help people discover this. The README (MSRV still says 1.80.0) and `docs/FEATURES.md` should also mention authorization.

File: src/v1/client.rs (hunk 19 above)
Comment (peterhuene) [reply src/v1/client.rs]: Ah yeah, I split out the `auth` module without coming back and fixing this.
